### ISSUE TITLE
[okta-react] autoRenew not renewing when it should

### ISSUE THREAD

**Issue by swiftone** [okta/okta-oidc-js#744]
@swiftone, no i don't have a separate issue (i can create one if you'd like). But again, i'm using the signin widget, along with okta-react 3.0, mobx, hash router.  Here's my configuration...
```
const CustomLoginRouter = () => {
	// console.log('CustomLoginRouter');

	const history = useHistory();

	const customAuthHandler = () => {
		console.log('customAuthHandler');
		history.push('/login');
	};

	return (
		<Security
			clientId="******************"
			issuer="https://************/oauth2/default"
			onAuthRequired={customAuthHandler}
			pkce
			redirectUri={`${window.location.origin}/implicit/callback`}
		>
			<Route
				component={Home}
				exact
				path="/"
			/>
			<Route
				component={LoginCallback}
				path="/implicit/callback"
			/>
			<Route
				component={Login}
				path="/login"
			/>
			<SecureRoute
				component={Protected}
				path="/Protected"
			/>
		</Security>
	);
};


const App = () => (
	<Router>
		<CustomLoginRouter />
	</Router>
);

export default App;
```
and also, here's how i have "Protected" component setup...
```
//export default (props) => {  // 	<------------------- I've tried this as well
export default withOktaAuth((props) => {
	const { authState, authService } = useOktaAuth();
	....
	useEffect(() => { //		<-------------------  never called after first render
		if (authState.isAuthenticated) {
			console.log('User is Authenticated!', authState);
		} else {
			console.log('User is NOT authenticated.', authState);
			logout();
		}
	}, [authState, authService]);
	....
});
```

_Originally posted by @sero323 in https://github.com/okta/okta-oidc-js/issues/721#issuecomment-612228180_

**Comment by swiftone** [comment]
Internal ref: OKTA-291520

**Comment by svantreeck** [comment]
For anyone having a similar issue to this, the way I'm working around it for now is calling `authService.updateAuthState()` on a timer, and if your tokens are out of date, it will update them if you still have a valid session.  Hopefully a fix will be implemented soon on the package, but its working out in a pinch for now

One caveat with this is you will need to set your Token Manager's `expireEarlySeconds` to a higher value than the frequency auth timer...the default value is 30s.  For my setup, my access tokens expire in 60 minutes, so I'm expiring them 5 minutes early and checking if the token is out of date every 2 minutes.

```ts
const config = {
  issuer: <myissuer>,
  clientId: <myclientid>,
  redirectUri: <myredirecturi>,
  tokenManager: {
    expireEarlySeconds: 300 // 5 minutes early
  }
}

const App = () => {
  return (<Security {...config}> 
   ..more app code...
 </Security>);
};


...somewhere in your code...
    const { authState, authService } = useOktaAuth();

    // check every two minutes if token is out of date
    useEffect(() => {
        const interval = setInterval(() => {
            if (authState.isAuthenticated) {
                authService.updateAuthState();
            }
        }, 1000 * 60 * 2);
        return () => clearInterval(interval);
    }, []);
```

**Comment by dusan-dragon** [comment]
@swiftone Could you also please explain this part from documentation: `When an expired token is requested by the library, a renewal request is executed to update the token.`

What do you mean by `requested by the library`?

I would expect that if the autoRenew is set to true (currently it is default, but not working) that it will do everything on its own and `useOktaAuth` hook will will trigger re-render of the component.

**Comment by hboylan** [comment]
@swiftone Any thoughts about where this issue might be coming from?

**Comment by swiftone** [comment]
@hboylan - Several thoughts, which we are actively investigating.

TL;DR: Make sure you've upgraded to the latest okta-auth-js (a dependency of okta-react) and keep watch for more updates

Full thoughts:

The code challenge that PKCE uses during authentication and to prevent MITM attacks is being lost before the process is complete, due to overlapping auth requests.  This would all be happening in the underlying okta-auth-js library that all our front end SDKs (including okta-react) are using.

Our first fix was to shift the (default) storage of this information from localStorage to sessionStorage, so multiple tabs doing auth because of expiration wouldn't stomp over each other.  That's in okta-auth-js 3.1.2, which older okta-react's will use on clean installs, and which the newest okta-react 3.0.2 requires.

However some users are still impacted, because their apps have overlapping auth requests even in a single tab.  The exact triggers for this will vary based on how your app is put together (which is why we've had a hard time reliably recreating - it's a race condition PLUS dependent on app structure).  

To address that we are considering short term and long term fixes.

Short term: We're working on preventing overlapping requests.
Long term: Making auth requests more explicit (rather than magic side-effects) and exposing the steps, including pushing and expanding some concepts done recently in okta-react about fewer async side-effects down to okta-auth-js.  

Those curious can take a peek at active PRs in okta-auth-js to see the work currently being undertaken, but we've prioritized this issue.

**Comment by hboylan** [comment]
@swiftone Thank you for the detailed explanation of the underlying issue.

In our case, overlapping requests were not the problem. Instead, the token refresh request is not called at the expected time of expiration. This caused our service client to continue using the expired token, since we relied on the `useOktaAuth` hook to get the token. 😞 
```typescript
export const GraphQLClient = ({ children }) => {
  const { authState } = useOktaAuth()

  const client = useMemo(
    () =>
      new ApolloClient({
        uri: process.env.REACT_APP_GRAPHQL_URL,
        cache: new InMemoryCache(),
        headers: {
          Authorization: `Bearer ${authState.token}`,
        },
      }),
    [authState.token],
  )

  return <ApolloProvider client={client}>{children}</ApolloProvider>
}
```

To solve this issue, we get the access token from token manager on every request. This method triggers the desired refresh/renewal if the token has expired. 😄 
```typescript
export const GraphQLClient = ({ children }) => {
  const { authService } = useOktaAuth()

  const client = useMemo(
    () =>
      new ApolloClient({
        link: from([
          setContext(async () => {
            const token = await authService.getAccessToken()
            return {
              headers: { Authorization: `Bearer ${token}` },
            }
          }),
          new HttpLink({ uri: process.env.REACT_APP_GRAPHQL_URL }),
        ]),
        cache: new InMemoryCache(),
      }),
    [authService],
  )

  return <ApolloProvider client={client}>{children}</ApolloProvider>
}
```

We find this method to be more reliable than @svantreeck's method of triggering `updateAuthState` at specific intervals. Hope this helps!

**Comment by veswill3** [comment]
@hboylan I am dealing with the exact same issue at the moment and also came to the conclusion: we switched from passing in `authState` to passing in the `authService` and calling `getAccessToken`, which seems to update the token automatically.
Just FYI to be careful on that use memo though. If they add something that updates the token in `authState`, you will end up creating a new apollo client (and thus a new InMemoryCache) which will effectively blow away you cache (I think)

**Comment by DavidLozzi** [comment]
Question related, the `expiresAt` value that I see in local storage is a long number, when I convert this to a date, I get a date/time in January of 1970. Examples I see from other support articles show an actual date/time in this value. What does this number represent? How do I know when it expires? Is it supposed to be a real date instead of the long number?

**Comment by hboylan** [comment]
The `expiresAt` number is a [Unix timestamp](https://www.unixtimestamp.com/index.php). You could use moment to convert:
```typescript
const getTokenExpiration = useCallback(() => {
  return authService
    .getTokenManager()
    .get('accessToken')
    .then(({ expiresAt }) => 
      moment.unix(expiresAt)
    )
}, [authService])
```

**Comment by aarongranick-okta** [comment]
For those encountering this issue, I hope this information will be helpful:
`@okta/okta-auth-js` 3.x and earlier use a **passive** form of autoRenew. This means that there will be no attempt to renew a token until it is **accessed** using `tokenManager.get()`. If `options.autoRenew` is true, then `tokenManager.get()` will automatically renew tokens and resolve the promise to a valid token unless the session has expired. 

Accessing the token from tokenManager before making the request, as @hboylan and @veswill3 are doing will perform the autoRenew logic. However, the token which is stored on the authState object exposed on the React context may be expired or out of date. This should be fixed in the latest release of `okta-react` (version 3.0.3) https://github.com/okta/okta-oidc-js/pull/826 so I think this workaround is no longer necessary: the token on authState **should** be current.

@svantreeck your method of calling on an interval works because `updateAuthState` will internally call `tokenManager.get()` which will perform the autoRenew logic. However, I think you would get better results by calling `updateAuthState` in response to the "expired" event which is emitted from tokenManager 30 seconds before the token is set to expire. This logic has now been implemented in `okta-react` (version 3.0.3) https://github.com/okta/okta-oidc-js/pull/826 so you should not need this workaround anymore.
